Source/AuthenticationInterceptor.swift

```diff
@@ -0,0 +1,405 @@
+    func isRequest(_ urlRequest: URLRequest, authenticatedWith credential: Credential) -> Bool
+}
+
+// MARK: -
+
+/// Represents various authentication failures that occur when using the `AuthenticationInterceptor`. All errors are
+/// still vended from Alamofire as `AFError` types. The `AuthenticationError` instances will be embedded within
+/// `AFError` `.requestAdaptationFailed` or `.requestRetryFailed` cases.
+public enum AuthenticationError: Error {
+    /// The credential was missing so the request could not be authenticated.
+    case missingCredential
+    /// The credential was refreshed too many times within the `RefreshWindow`.
+    case excessiveRefresh
+}
+
+// MARK: -
+
+/// The `AuthenticationInterceptor` class manages the queuing and threading complexity of authenticating requests.
+/// It relies on an `Authenticator` type to handle the actual `URLRequest` authentication and `Credential` refresh.
+public class AuthenticationInterceptor<AuthenticatorType>: RequestInterceptor where AuthenticatorType: Authenticator {
+
+    // MARK: Typealiases
+
+    /// Type of credential used to authenticate requests.
+    public typealias Credential = AuthenticatorType.Credential
+
+    // MARK: Helper Types
+
+    /// Type that defines a time window used to identify excessive refresh calls. When enabled, prior to executing a
+    /// refresh, the `AuthenticationInterceptor` compares the timestamp history of previous refresh calls against the
+    /// `RefreshWindow`. If more refreshes have occurred within the refresh window than allowed, the refresh is
+    /// cancelled and an `AuthorizationError.excessiveRefresh` error is thrown.
+    public struct RefreshWindow {
+        /// `TimeInterval` defining the duration of the time window before the current time in which the number of
+        /// refresh attempts is compared against `maximumAttempts`. For example, if `interval` is 30 seconds, then the
+        /// `RefreshWindow` represents the past 30 seconds. If more attempts occurred in the past 30 seconds than
+        /// `maximumAttempts`, an `.excessiveRefresh` error will be thrown.
+        public let interval: TimeInterval
+
+        /// Total refresh attempts allowed within `interval` before throwing an `.excessiveRefresh` error.
+        public let maximumAttempts: Int
+
+        /// Creates a `RefreshWindow` instance from the specified `interval` and `maximumAttempts`.
+        ///
+        /// - Parameters:
+        ///   - interval:        `TimeInterval` defining the duration of the time window before the current time.
+        ///   - maximumAttempts: The maximum attempts allowed within the `TimeInterval`.
+        public init(interval: TimeInterval = 30.0, maximumAttempts: Int = 5) {
+            self.interval = interval
+            self.maximumAttempts = maximumAttempts
+        }
+    }
+
+    private struct AdaptOperation {
+        let urlRequest: URLRequest
+        let session: Alamofire.Session
```

Comment:
Qualified `Alamofire.` seem unnecessary here.

Reply:
Fixed in 107aba21.